Update documentation.

dan-zheng opened the pull request:
Add doc comments for:
- `Structural` protocol
- `StructuralStruct`, `StructuralProperty`
- `StructuralEnum`, `StructuralCase`
- `StructuralCons`, `StructuralEmpty`, `StructuralEither`

Rename properties:
- `StructuralStruct.type` -> `StructuralStruct.representedType`
- `StructuralStruct.properties` -> `StructuralStruct.body`
- `StructuralEnum.type` -> `StructuralEnum.representedType`
- `StructuralEnum.cases` -> `StructuralEnum.body`

---

Documentation changes were discussed during Swift for TensorFlow team meeting.
There's further room for improvement, we can discuss on this PR.

dan-zheng committed:
Update documentation.

Add doc comments for:
- `Structural` protocol
- `StructuralStruct`, `StructuralProperty`
- `StructuralEnum`, `StructuralCase`
- `StructuralCons`, `StructuralEmpty`, `StructuralEither`

Rename properties:
- `StructuralStruct.type` -> `StructuralStruct.representedType`
- `StructuralStruct.properties` -> `StructuralStruct.body`
- `StructuralEnum.type` -> `StructuralEnum.representedType`
- `StructuralEnum.cases` -> `StructuralEnum.body`

dan-zheng commented on Sources/StructuralCore/Structural.swift:
```diff
@@ -12,7 +12,60 @@
+///     extension Color: Structural {
+///         // Compiler synthesizes:
+///         typealias StructuralRepresentation =
+///             StructuralEnum<
```
`StructuralCons` and `StructuralEither` are both type-level heterogeneous lists.

`StructuralCons` (for struct stored properties and enum associated values) ends with a `StructuralEmpty` empty list base case.

However, `StructuralEither` seems to have no such base case. **Should it (also use `StructuralEmpty`)?**

This seems necessary for defining `StructuralRepresentation` for an empty enum (`StructuralEnum<StructuralEmpty>`).

shabalind replied:
Thanks for pointing this out, it's indeed seems to be inconsistent! Feel free to address this in a follow-up PR (examples protocols & structurs might have to be updated to fully validate this direction).

dan-zheng commented on Sources/StructuralCore/Structural.swift:
```diff
@@ -21,103 +74,187 @@ public protocol Structural {
 public struct StructuralStruct<Properties> {
-    public var type: Any.Type?
-    public var properties: Properties
+    /// The type of the represented struct type.
```
Note: it is not easy to explain the case where `representedType` can be `nil`:
```
The type of the represented struct type, or `nil` if ...
```
This may be a design smell.

shabalind replied:
Yep, as discussed earlier it might be desirable to address this. Similarly to the other comment, it's better to do it in a dedicated PR to validate changes over the uses of previous encoding that allowed for `nil` type.

dan-zheng commented:
CI fails because I have to update the usages of renamed declarations (e.g. `type` → `representedType`).
I will do that later today.

dan-zheng commented on Sources/StructuralCore/Structural.swift:
```diff
@@ -21,103 +74,187 @@ public protocol Structural {
-    public var type: Any.Type?
-    public var properties: Properties
+    /// The type of the represented struct type.
+    public var representedType: Any.Type?
+    /// The names and values of stored properties.
+    public var body: Properties
```
I wonder why all properties in this file are `var` instead of `let`? Are there use cases for `var`?

shabalind replied:
The use case is to allow easy in-place mutation of the `StructuralRepresentation` (see `MyInplaceAdd` protocol for an example). Otherwise if make everything immutable, users will have to do a deep copy of the whole representation every time you change a single part of it.

dan-zheng replied:
I see, thanks!

I think in-place mutation is a good reason for making `StructuralStruct.body` and `StructuralEnum.body` be mutable.

How about `var StructuralProperty.name` (mutable property name) and `var StructuralProperty.isMutable` (mutable property mutability) - are there use cases for those? Even if not, keeping those as `var` sounds good to me for now, in case there are future use cases.

shabalind replied:
Changes to `name` and `isMutable` would not have any effect, so it makes sense to make them `let` instead.

shabalind commented on Sources/StructuralCore/Structural.swift:
```diff
@@ -21,103 +74,187 @@ public protocol Structural {
+/// - Note: `StructuralEmpty<...>` is used as the base case of the type-level
+///   heterogeneous cons list `StructuralCons`. If variadiac generics are added
+///   to the Swift language, `StructuralCons` can be deprecated in favor of a
+///   less-verbose heterogeneous list representation using variadic generics.
```
I'm not sure if mentioning the variadic generics is relavant, given that there is no concrete proposal at the moment (E.g.: what if variadic generics are compiled down to heterogeneous lists?).

dan-zheng replied:
Removing mentions of variadic generics sounds good to me! I removed them for now.

> (E.g.: what if variadic generics are compiled down to heterogeneous lists?).

Minor comment: if variadic generics are added, I think what would matter is their user-facing API (enabling a simpler `StructuralStruct<...>` design), not how they are compiled down.

shabalind commented on Sources/StructuralCore/Structural.swift:
```diff
@@ -21,103 +74,187 @@ public protocol Structural {
+///   list, which is verbose and not particularly human-readable. If variadiac
+///   generics are added to the Swift language, `StructuralEither` can be
+///   deprecated in favor of a less-verbose heterogeneous list representaiton
+///   using variadic generics.
```
Same comment regarding variadic generics as above.

dan-zheng replied:
Removed for now.

shabalind reviewed (approved):
Overall LGTM, some minor comments below.

dan-zheng committed:
Remove details about variadic generics.

Address review feedback.
There is no accepted proposal yet adding variadic generics to Swift.

dan-zheng commented:
I'm merging this PR now to unblock progress. Happy to address further feedback later!
